apollo-router/src/plugins/traffic_shaping/mod.rs

```diff
@@ -498,9 +498,9 @@ mod test {
             .oneshot(SubgraphRequest::fake_builder().build())
             .await
             .unwrap();
-        // Note: use `timeout` to guarantee 300ms has elapsed
+        // Note: use `timeout` to guarantee 400ms has elapsed
         let big_sleep = tokio::time::sleep(Duration::from_secs(10));
-        assert!(tokio::time::timeout(Duration::from_millis(300), big_sleep)
+        assert!(tokio::time::timeout(Duration::from_millis(400), big_sleep)
             .await
             .is_err());
```

Comment:
By the way, how does `timeout` provide any more guarantee than plain `sleep`?

Reply:
I saw something in the tokio docs that implied a sleep may finish early due to granularity (resolution) of duration, whereas there were no comments on the timeout function. However, in retrospect, that may be just a side effect of the documentation writers rather than an actual change in behaviour.

I don't have actual numbers, but it does seem like timeout has reduced the number of failures, but that could be just wishful thinking.

The downside with setting the timeout higher (400ms) is that it invalidates the test. The test is supposed to be measuring rate limiting and now we aren't.

Perhaps a better approach would be to repeat the test within a test a number of times and call it a pass if a majority pass? Something like 3 out of 5 or whatever we deem to be statistically significant.

Reply:
How about running `sleep` in a loop until the test measures that 300ms have actually passed?